File: Python3/807. Max Increase to Keep City Skyline.py
class Solution:
    def maxIncreaseKeepingSkyline(self, grid: List[List[int]]) -> int:
        #n = len(grid)
        #xmax = [max(grid[i]) for i in range(n)]
        #ymax = [max([grid[i][j] for i in range(n)]) for j in range(n)]
        #return sum(min(xmax[i], ymax[j]) - grid[i][j] for i in range(n) for j in range(n))

        xmax = list(map(max, grid))
        ymax = list(map(max, zip(*grid)))
        return sum(min(xmax[i], ymax[j]) - h for i, row in enumerate(grid) for j, h in enumerate(row))

File: C++/807. Max Increase to Keep City Skyline.cpp
class Solution {
public:
    int maxIncreaseKeepingSkyline(vector<vector<int>>& grid) {
        int n = grid.size();
        int res = 0;
        vector<int> xmax(n, 0);
        vector<int> ymax(n, 0);
        for(int i=0;i<n;i++) {
            xmax[i] = *max_element(grid[i].begin(), grid[i].end());
        }
        for(int j=0;j<n;j++) {
            for(int i=0;i<n;i++) {
                ymax[j] = max(ymax[j], grid[i][j]);
            }
            for(int i=0;i<n;i++) {
                res += min(xmax[i], ymax[j]) - grid[i][j];
            }
        }
        return res;
    }
};
